adds CONTRIBUTING document to the main Swift repository (#75654)
swiftlang/swift: CONTRIBUTING.md

## Reporting Bugs

Reporting bugs is a great way for anyone to help improve Swift.
The open source Swift project uses GitHub Issues for tracking bugs.

If a bug can be reproduced only within an Xcode project or a playground, or if the bug is associated with an Apple NDA, please file a report to Apple’s [bug reporter](https://bugreport.apple.com/) instead.
When opening an issue, please include the following:

- **A concise description of the problem**.
If the issue is a crash, include a stack trace.
Otherwise, describe the behavior you were expecting to see, along with the behavior you actually observed.

- **A reproducible test case**.
Double-check that your test case reproduces the issue.
A relatively small sample (roughly within 50 lines of code) is best pasted directly into the description; a larger one may be uploaded as an attachment.
Consider reducing the sample to the smallest amount of code possible; a smaller test case is easier to reason about and more appealing to сontributors.

- **A description of the environment that reproduces the problem**.

> Include information about the Swift compiler’s version, the deployment target (if explicitly set) and your platform.

Author: It would probably be better as a github issue template as well

> Include information about the Swift compiler’s version, the deployment target (if explicitly set) and your platform.

Reviewer: That’s what I meant: We have GitHub issue forms that ask for all this information individually.

Good first issues are decorated with a corresponding label and are most easily found by visiting `github.com/apple/<repository>/contribute`.
An example for the main Swift repository is [github.com/apple/swiftlang/contribute](https://github.com/swiftlang/swift/contribute).
They are expected to be low-priority and of modest scope, and not require excessive refactoring, research, or debugging — rather, they should encourage newcomers to dip their toes in some part of Swift, learn more about it, and make a real contribution.

Anyone with [commit access](https://www.swift.org/contributing/#commit-access) and insight into a particular area is welcome and encouraged to pin down or think up good first issues.

## Contributing Code

If you are interested in:

- Contributing fixes and features to the compiler: See our
  [How to Submit Your First Pull Request guide](/docs/HowToGuides/FirstPullRequest.md).
- Building the compiler as a one-off: See our [Getting Started guide][].
- Building a toolchain as a one-off: Follow the [Getting Started guide][]
  up until the "Building the project" section. After that, follow the

>   instructions in the [Swift Toolchains](/README.md#swift-toolchains).

Reviewer: I don’t think that a new contributor would want to build an entire toolchain. Doing so is its own can of worms and it’s rarely required. So I’d just remove the bullet. (You should be able to inline the Getting started guide link again as well.

Author: removed bullet and inlined the link to the getting started docs

> We also host answers to [Frequently Aasked Questions](/docs/HowToGuides/FAQ.md) that may be of interest.

Reviewer: ```suggestion
We also host answers to [Frequently Asked Questions](/docs/HowToGuides/FAQ.md) that may be of interest.
```

- Resolving merge conflicts can take a lot of time if branch development and mainline development occur in the same pieces of code.
- People in the community tend to ignore work on branches.
- Very large changes are difficult to code review.
- Branches are not routinely tested by the continuous integration infrastructure.

To address these problems, Swift uses an incremental development style. Small changes are preferred whenever possible.
We require contributors to follow this practice when making large or otherwise invasive changes. Some tips follow:

- Large or invasive changes usually have secondary changes that must be made before the large change (for example, API cleanup or addition).
Commit these changes before the major change, independently of that work.

- If possible, decompose the remaining interrelated work into unrelated sets of changes.
Next, define the first increment and get consensus on the development goal of the change.

- Make each change in the set either stand alone (for example, to fix a bug) or part of a planned series of changes that work toward the development goal.
Explaining these relationships to the community can be helpful.

> If you are interested in making a large change and feel unsure about its overall effect, please make sure to first discuss the change and reach a consensus through the developer forums.

Reviewer: ```suggestion
If you are interested in making a large change and feel unsure about its overall effect, please make sure to first discuss the change and reach a consensus through the https://forums.swift.org.
```

Author: make this a link, not just a raw URL

- In changes that are restricted to a specific part of the code, include a [tag] at the start of the line in square brackets—for example, “[stdlib] …” or “[SILGen] …”.
This tag helps email filters and searches for post-commit reviews.
- When there is a body, separate it from the title by an empty line.
- Make body concise, while including the complete reasoning.
Unless required to understand the change, additional code examples or other details should be left to bug comments or the mailing list.
- If the commit fixes an issue in the bug tracking system, include a link to the issue in the message.
- For text formatting and spelling, follow the same rules as documentation and in-code comments—for example, the use of capitalization and periods.
- If the commit is a bug fix on top of another recently committed change, or a revert or reapply of a patch, include the Git revision number of the prior related commit, for example “Revert abcdef because it caused bug#”.

For minor violations of these guidelines, the community normally favors reminding the contributor of this policy over reverting.

> Minor corrections and omissions can be handled by sending a reply to the commits mailing list.

Reviewer: I have no idea what this sentence is supposed to refer to – we don’t have a commits mailing list and if we do there is no significant traffic on it. Let’s delete it.

Author: dropping both sentences

### Attribution of Changes

When contributors submit a change to a Swift subproject, after that change is approved, other developers with commit access may commit it for the author.
When doing so, it is important to retain correct attribution of the contribution. Generally speaking, Git handles attribution automatically.

We do not want the source code to be littered with random attributions like “this code written by J. Random Hacker”, which is noisy and distracting.
Do not add contributor names to the source code or documentation.

In addition, don’t commit changes authored by others unless they have submitted the change to the project or you have been authorized to submit on their behalf—for example, you work together and your company authorized you to contribute the changes.
The author should first either submit the change through a pull request to the relevant project, email the development list, or add a bug tracker item.

> If someone sends you a change privately, encourage them to submit it to the appropriate list first.

Reviewer: I don’t know how ancient this paragraph it but most of it isn’t relevant with git and GitHub. I think the only relevant part is

```md
Don’t commit changes authored by others unless they have submitted the change to the project or you have been authorized to submit on their behalf — for example, you work together and your company authorized you to contribute the changes.
```

Author: removed the non-github pieces and left in the "don't submit the work of others" bit with the caveat

### Code Templates

Code Templates

> As mentioned in the Community Overview, the license and copyright protections for Swift.org code are called out at the top of every source code file.

Reviewer: There is no *Community Overview* section in this document.

```suggestion
The license and copyright protections for Swift.org code are called out at the top of every source code file.
```

> On the rare occasion you contribute a change that includes a new source file, ensure that the header is filled out appropriately.

Reviewer: I don’t think it’s that rare to contribute a new source file…

```suggestion
If you contribute a change that includes a new source file, ensure that the header is filled out appropriately.
```

//
// Copyright (c) 2024 Apple Inc. and the Swift project authors
// Licensed under Apache License v2.0 with Runtime Library Exception
//
// See https://swift.org/LICENSE.txt for license information
// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
//
//===----------------------------------------------------------------------===//
///
/// \file
/// This file contains stuff that I am describing here in the header and will
/// be sure to keep up to date.
///
//===----------------------------------------------------------------------===//
```

The divider lines should be exactly 80 characters wide to aid in adherence to the code style guidelines.
The bottom section contains an optional description intended for generated documentation (these lines begin with `///` rather than `//`).
If there is no description, this area can be skipped.

### Code Review

The Swift project relies heavily on code review to improve software quality:

- All significant changes, by all developers, must be reviewed before they are committed to the repository.
Smaller changes (or changes where the developer owns the component) can be reviewed after being committed.

> - Code reviews are conducted on GitHub (through comments on pull requests or commits) and are reflected on the relevant project’s commit mailing list.

Reviewer: ```suggestion
- Code reviews are conducted on GitHub.
```

- The developer responsible for a code change is also responsible for making all necessary review-related changes.

> Code review can be an iterative process, which continues until the change is ready to be committed.

Reviewer: ```suggestion
Code review can be an iterative process, which continues until the change is ready to be merged.
```

Note that anyone is welcome to review and give feedback on a change, but only people with commit access to the repository can approve it.

### Testing

Developers are required to create test cases for any bugs fixed and any new features added, and to contribute them along with the changes.

- All feature and regression test cases are added to the appropriate test directory—for example, the `swift/test` directory.
- Write test cases at the abstraction level nearest to the actual feature.
For example, if it’s a Swift language feature, write it in Swift; if it’s a SIL optimization, write it in SIL.
- Reduce test cases as much as possible, especially for regressions.
It’s unacceptable to place an entire failing program into `swift/test` because this slows down testing for all developers.
Please keep them short.

### Quality

People depend on Swift to create their production software.
This means that a bug in Swift could cause bugs in thousands, even millions of developers’ products.
Because of this, the Swift project maintains a high bar for quality.
The minimum quality standards that any change must satisfy before being committed to the main development branch include:

> 1. Code must compile without errors or warnings on at least one platform.

Reviewer: ```suggestion
1. Code must compile without errors on all supported platforms and should be free of warnings on at least one platform.
```

3. Code must pass the appropriate test suites—for example, the `swift/test` and `swift/validation-test` test suites in the Swift compiler.

Additionally, the committer is responsible for addressing any problems found in the future that the change may cause.
This responsibility means that you may need to update your change in order to:

- Ensure the code compiles cleanly on all primary platforms.
- Fix any correctness regressions found in other test suites.
- Fix any major performance regressions.
- Fix any performance or correctness regressions in the downstream Swift tools.
- Fix any performance or correctness regressions that result in customer code that uses Swift.
- Address any bugs that appear in the bug tracker as a result from your change.

We prefer that these issues be handled before submission, but we understand that it isn’t possible to test all of this for every submission.
Our continuous integration (CI) infrastructure normally finds these problems.
We recommend watching the CI infrastructure throughout the next day to look for regressions.
The CI infrastructure will directly email you if a group of commits that included yours caused a failure.

> You are expected to check those messages to see whether they are your fault and, if so, fix the breakage.

Reviewer: I don’t think this paragraph contributes to anything. CI testing is required for a PR to be merged and I don’t think it’s reasonable for anyone to *watch the CI infrastrucute* throughout the next day. Also, the CI infrastructure doesn’t send failure emails. Let’s just remove it.

Author: Should be gone now with the latest commit

Commit access is granted to contributors with a track record of submitting high-quality changes.
If you would like commit access, please send an email to [the code owners list](code-owners@forums.swift.org) with the GitHub user name that you want to use and a list of 5 non-trivial pull requests that were accepted without modifications.

Once you’ve been granted commit access, you will be able to commit to all of the GitHub repositories that host Swift.org projects.
To verify that your commit access works, please make a test commit (for example, change a comment or add a blank line). The following policies apply to users with commit access:

- You are granted commit-after-approval to all parts of Swift.
To get approval, create a pull request. When the pull request is approved, you may merge it yourself.

- You may commit an obvious change without first getting approval.
The community expects you to use good judgment.
Examples are reverting obviously broken patches, correcting code comments, and other minor changes.

- You are allowed to commit changes without approval to the portions of Swift to which you have contributed or for which you have been assigned responsibility.

> Such commits must not break the build.

Reviewer: CI testing prevents you from breaking the build. Let’s remove this.

Author: Should be gone now with the latest commit

This repository track upstreams LLVM development and contains additional changes for Swift.
The upstream LLVM repository are merged into the Swift-specific repository frequently.
Every attempt is made to minimize the differences between upstream LLVM and the Apple fork to only those changes specifically required for Swift.

### Where Do LLVM Changes Go?

Swift follows a policy of making a change in the most upstream repository that is feasible.
Contributions to Swift that involve Apple’s version of LLVM Project should go directly into the upstream LLVM repository unless they are specific to Swift.
For example, an improvement to LLDB’s data formatters for a Swift type belongs in the Apple LLVM Project repository, whereas a bug fix to an LLVM optimizer—even if it’s only been observed when operating on Swift-generated LLVM IR—belongs in upstream LLVM.

Commits to an upstream LLVM repository are automatically merged into the appropriate upstream branches in the corresponding Swift repository (`next`) in the [llvm-project](https://github.com/apple/llvm-project).

### Swift and LLVM Developer Policies

> Contributions to Swift’s LLVM or Clang clones are governed by the [LLVM Developer Policy](http://llvm.org/docs/DeveloperPolicy.html) and should follow the appropriate [licensing](http://llvm.org/docs/DeveloperPolicy.html#copyright-license-and-patents) and [coding standards](http://llvm.org/docs/CodingStandards.html).

Reviewer: ```suggestion
Contributions to [llvm-project clone](https://github.com/swiftlang/llvm-project) are governed by the [LLVM Developer Policy](http://llvm.org/docs/DeveloperPolicy.html) and should follow the appropriate [licensing](http://llvm.org/docs/DeveloperPolicy.html#copyright-license-and-patents) and [coding standards](http://llvm.org/docs/CodingStandards.html).
```
